woocommerce/Lifecycle.php

```diff
@@ -0,0 +1,292 @@
+
+	/** @var \SkyVerge\WooCommerce\PluginFramework\v5_0_1\SV_WC_Plugin plugin instance */
+	private $plugin;
+
+
+	/**
+	 * Constructs the class.
+	 *
+	 * @since 5.1.0-dev
+	 *
+	 * @param \SkyVerge\WooCommerce\PluginFramework\v5_0_1\SV_WC_Plugin $plugin plugin instance
+	 */
+	public function __construct( \SkyVerge\WooCommerce\PluginFramework\v5_0_1\SV_WC_Plugin $plugin ) {
+
+		$this->plugin = $plugin;
+
+		$this->add_hooks();
+	}
+
+
+	/**
+	 * Adds the action & filter hooks.
+	 *
+	 * @since 5.1.0-dev
+	 */
+	protected function add_hooks() {
+
+		add_action( 'wc_' . $this->get_plugin()->get_id() . '_updated', array( $this, 'do_update' ) );
+
+		add_action( 'init', array( $this, 'add_admin_notices' ) );
+
+		add_action( 'wc_' . $this->get_plugin()->get_id() . '_milestone_reached', array( $this, 'trigger_milestone' ), 10, 3 );
+	}
+
+
+	/**
+	 * Handles tasks after the plugin has been updated.
+	 *
+	 * @internal
+	 *
+	 * @since 5.1.0-dev
+	 */
+	public function do_update() {
+
+		// if the plugin never had any previous milestones, consider them all reached so their notices aren't displayed
+		if ( ! $this->get_milestone_version() ) {
+			$this->set_milestone_version( $this->get_plugin()->get_version() );
+		}
+	}
+
+
+	/**
+	 * Adds any lifecycle admin notices.
+	 *
+	 * @since 5.1.0-dev
+	 */
+	public function add_admin_notices() {
+
+		// display any milestone notices
+		foreach ( $this->get_milestone_messages() as $id => $message ) {
+
+			// bail if this notice was already dismissed
+			if ( ! $this->get_plugin()->get_admin_notice_handler()->should_display_notice( $id ) ) {
+				continue;
+			}
+
+			/**
+			 * Filters a milestone notice message.
+			 *
+			 * @since 5.1.0-dev
+			 *
+			 * @param string $message message text to be used for the milestone notice
+			 * @param string $id milestone ID
+			 */
+			$message = apply_filters( 'wc_' . $this->get_plugin()->get_id() . '_milestone_message', $this->generate_milestone_notice_message( $message ), $id );
+
+			if ( $message ) {
+
+				$this->get_plugin()->get_admin_notice_handler()->add_admin_notice( $message, $id );
+
+				// only display one notice at a time
+				break;
+			}
+		}
+	}
+
+
+	/** Milestone Methods *****************************************************/
+
+
+	/**
+	 * Triggers a milestone.
+	 *
+	 * This will only be triggered if the install's "milestone version" is lower
+	 * than $since. Plugins can specify $since as the version at which a
+	 * milestone's feature was added. This prevents existing installs from
+	 * triggering notices for milestones that have long passed, like a payment
+	 * gateway's first successful payment. Omitting $since will assume the
+	 * milestone has always existed and should only trigger for fresh installs.
+	 *
+	 * @since 5.1.0-dev
+	 *
+	 * @param string $id milestone ID
+	 * @param string $message message to display to the user
+	 * @param string $since the version since this milestone has existed in the plugin
+	 */
```

Comment:
missing return tag - this method seems to return false on version compare failure or void if sanity check passes

should it return something or bool? if nothing, can be just `return;`